File: packages/observing-api/src/routes/taxonomy.ts
/**
 * Taxonomy routes - search, validate, and taxon details
 */

import { Router } from "express";
import { Database, TaxonomyClient } from "observing-shared";
import { enrichOccurrences } from "../enrichment.js";
import { logger } from "../middleware/logging.js";

export function createTaxonomyRoutes(
  db: Database,
  taxonomy: TaxonomyClient
): Router {
  const router = Router();

  // Search taxa
  router.get("/search", async (req, res) => {
    try {
      const query = req.query["q"] as string;
      if (!query || query.length < 2) {
        res.status(400).json({ error: "Query must be at least 2 characters" });
        return;
      }

      const results = await taxonomy.search(query);
      res.json({ results });
    } catch (error) {
      logger.error({ err: error }, "Error searching taxa");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Validate taxon name
  router.get("/validate", async (req, res) => {
    try {
      const name = req.query["name"] as string;
      if (!name) {
        res.status(400).json({ error: "name parameter required" });
        return;
      }

      const result = await taxonomy.validate(name);
      res.json(result);
    } catch (error) {
      logger.error({ err: error }, "Error validating taxon");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Helper to resolve a taxon from an ID or name string
  async function resolveTaxon(idOrName: string) {
    if (idOrName.startsWith("gbif:")) {
      return taxonomy.getById(idOrName);
    }
    // Try name-based resolution
    return taxonomy.getByName(idOrName);
  }

  // Helper to resolve a taxon from kingdom + name
  async function resolveTaxonByKingdomName(kingdom: string, name: string) {
    return taxonomy.getByName(name, kingdom);
  }

  // Get taxon occurrences by kingdom + name (must be before /:param1/:param2)
  router.get("/:kingdom/:name/occurrences", async (req, res) => {
    try {
      const kingdom = decodeURIComponent(req.params["kingdom"] ?? "");
      const name = decodeURIComponent(req.params["name"] ?? "");
      const cursor = req.query["cursor"] as string | undefined;
      const limit = Math.min(parseInt(req.query["limit"] as string) || 20, 100);

      const taxon = await resolveTaxonByKingdomName(kingdom, name);
      if (!taxon) {
        res.status(404).json({ error: "Taxon not found" });
        return;
      }

      const rows = await db.getOccurrencesByTaxon(taxon.scientificName, taxon.rank, {
        limit,
        ...(cursor && { cursor }),
        ...(taxon.kingdom && { kingdom: taxon.kingdom }),
      });

      const occurrences = await enrichOccurrences(db, rows);

      const nextCursor =
        rows.length === limit
          ? rows[rows.length - 1]?.created_at?.toISOString()
          : undefined;

      res.json({
        occurrences,
        cursor: nextCursor,
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching taxon occurrences");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get taxon occurrences by ID or name (backward compat)
  router.get("/:id/occurrences", async (req, res) => {
    try {
      const idOrName = decodeURIComponent(req.params["id"] ?? "");
      const cursor = req.query["cursor"] as string | undefined;
      const limit = Math.min(parseInt(req.query["limit"] as string) || 20, 100);

      const taxon = await resolveTaxon(idOrName);
      if (!taxon) {
        res.status(404).json({ error: "Taxon not found" });
        return;
      }

      const rows = await db.getOccurrencesByTaxon(taxon.scientificName, taxon.rank, {
        limit,
        ...(cursor && { cursor }),
        ...(taxon.kingdom && { kingdom: taxon.kingdom }),
      });

      const occurrences = await enrichOccurrences(db, rows);

      const nextCursor =
        rows.length === limit
          ? rows[rows.length - 1]?.created_at?.toISOString()
          : undefined;

      res.json({
        occurrences,
        cursor: nextCursor,
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching taxon occurrences");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get taxon details by kingdom + name
  router.get("/:kingdom/:name", async (req, res) => {
    try {
      const kingdom = decodeURIComponent(req.params["kingdom"] ?? "");
      const name = decodeURIComponent(req.params["name"] ?? "");

      const taxon = await resolveTaxonByKingdomName(kingdom, name);
      if (!taxon) {
        res.status(404).json({ error: "Taxon not found" });
        return;
      }

      const observationCount = await db.countOccurrencesByTaxon(
        taxon.scientificName,
        taxon.rank,
        taxon.kingdom,
      );

      res.json({
        ...taxon,
        observationCount,
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching taxon");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Get taxon details by ID or name (backward compat + kingdom-level lookup)
  router.get("/:id", async (req, res) => {
    try {
      const idOrName = decodeURIComponent(req.params["id"] ?? "");

      const taxon = await resolveTaxon(idOrName);
      if (!taxon) {
        res.status(404).json({ error: "Taxon not found" });
        return;
      }

      const observationCount = await db.countOccurrencesByTaxon(
        taxon.scientificName,
        taxon.rank,
        taxon.kingdom,
      );

      res.json({
        ...taxon,
        observationCount,
      });
    } catch (error) {
      logger.error({ err: error }, "Error fetching taxon");
      res.status(500).json({ error: "Internal server error" });
    }
  });

  return router;
}

File: crates/observing-appview/src/routes/taxonomy.rs
use axum::extract::{Path, Query, State};
use axum::Json;
use observing_db::types::TaxonOccurrenceOptions;
use serde::Deserialize;

use crate::auth::session_did;
use crate::constants;
use crate::enrichment;
use crate::error::AppError;
use crate::responses::{OccurrenceListResponse, TaxonSearchResponse};
use crate::state::AppState;
use crate::taxonomy_client::{TaxonDetailWithCount, ValidateResponse};

#[derive(Deserialize)]
pub struct SearchParams {
    q: Option<String>,
}

pub async fn search(
    State(state): State<AppState>,
    Query(params): Query<SearchParams>,
) -> Result<Json<TaxonSearchResponse>, AppError> {
    let query = params
        .q
        .ok_or_else(|| AppError::BadRequest("q is required".into()))?;

    if query.len() < constants::MIN_SEARCH_QUERY_LENGTH {
        return Err(AppError::BadRequest(format!(
            "Search query must be at least {} characters",
            constants::MIN_SEARCH_QUERY_LENGTH
        )));
    }

    let results = state
        .taxonomy
        .search(&query, None)
        .await
        .unwrap_or_default();

    Ok(Json(TaxonSearchResponse { results }))
}

#[derive(Deserialize)]
pub struct ValidateParams {
    name: Option<String>,
}

pub async fn validate(
    State(state): State<AppState>,
    Query(params): Query<ValidateParams>,
) -> Result<Json<ValidateResponse>, AppError> {
    let name = params
        .name
        .ok_or_else(|| AppError::BadRequest("name is required".into()))?;

    match state.taxonomy.validate(&name).await {
        Some(result) => Ok(Json(result)),
        None => Ok(Json(ValidateResponse {
            valid: false,
            matched_name: None,
            taxon: None,
            suggestions: Some(vec![]),
        })),
    }
}

#[derive(Deserialize)]
pub struct TaxonOccurrenceParams {
    limit: Option<i64>,
    cursor: Option<String>,
}

pub async fn get_taxon_by_kingdom_name(
    State(state): State<AppState>,
    Path((kingdom, name)): Path<(String, String)>,
) -> Result<Json<TaxonDetailWithCount>, AppError> {
    // Frontend uses dashes in URLs (e.g., "Morus-alba"), convert to spaces
    let name = name.replace('-', " ");

    let detail = state
        .taxonomy
        .get_by_name(&name, Some(&kingdom))
        .await?
        .ok_or_else(|| AppError::NotFound("Taxon not found".into()))?;

    let count = observing_db::feeds::count_occurrences_by_taxon(
        &state.pool,
        &name,
        &detail.rank,
        Some(&kingdom),
    )
    .await
    .unwrap_or(0);

    Ok(Json(TaxonDetailWithCount {
        detail,
        observation_count: count,
    }))
}

pub async fn get_children_by_kingdom_name(
    State(state): State<AppState>,
    Path((kingdom, name)): Path<(String, String)>,
) -> Result<Json<Vec<crate::taxonomy_client::TaxonResult>>, AppError> {
    let name = name.replace('-', " ");
    let children = state
        .taxonomy
        .get_children(&name, Some(&kingdom))
        .await
        .unwrap_or(None)
        .unwrap_or_default();
    Ok(Json(children))
}

pub async fn get_taxon_occurrences_by_kingdom_name(
    State(state): State<AppState>,
    cookies: axum_extra::extract::CookieJar,
    Path((kingdom, name)): Path<(String, String)>,
    Query(params): Query<TaxonOccurrenceParams>,
) -> Result<Json<OccurrenceListResponse>, AppError> {
    let limit = params
        .limit
        .unwrap_or(constants::DEFAULT_FEED_LIMIT)
        .min(constants::MAX_FEED_LIMIT);
    let name = name.replace('-', " ");

    // Look up taxon to get rank
    let detail = state
        .taxonomy
        .get_by_name(&name, Some(&kingdom))
        .await
        .unwrap_or(None);
    let rank = detail
        .as_ref()
        .map(|d| d.rank.clone())
        .unwrap_or_else(|| "species".to_string());

    let options = TaxonOccurrenceOptions {
        limit: Some(limit),
        cursor: params.cursor,
        kingdom: Some(kingdom),
    };

    let rows = observing_db::feeds::get_occurrences_by_taxon(
        &state.pool,
        &name,
        &rank,
        &options,
        &state.hidden_dids,
    )
    .await?;

    let viewer = session_did(&cookies);
    let occurrences = enrichment::enrich_occurrences(
        &state.pool,
        &state.resolver,
        &state.taxonomy,
        &rows,
        viewer.as_deref(),
    )
    .await;

    let next_cursor = occurrences.last().map(|o| o.created_at.clone());

    Ok(Json(OccurrenceListResponse {
        occurrences,
        cursor: next_cursor,
    }))
}

pub async fn get_taxon_by_id(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<TaxonDetailWithCount>, AppError> {
    let detail = state
        .taxonomy
        .get_by_id(&id)
        .await?
        .ok_or_else(|| AppError::NotFound("Taxon not found".into()))?;

    let count = observing_db::feeds::count_occurrences_by_taxon(
        &state.pool,
        &detail.scientific_name,
        &detail.rank,
        detail.kingdom.as_deref(),
    )
    .await
    .unwrap_or(0);

    Ok(Json(TaxonDetailWithCount {
        detail,
        observation_count: count,
    }))
}

pub async fn get_taxon_occurrences_by_id(
    State(state): State<AppState>,
    cookies: axum_extra::extract::CookieJar,
    Path(id): Path<String>,
    Query(params): Query<TaxonOccurrenceParams>,
) -> Result<Json<OccurrenceListResponse>, AppError> {
    let limit = params
        .limit
        .unwrap_or(constants::DEFAULT_FEED_LIMIT)
        .min(constants::MAX_FEED_LIMIT);

    // Look up taxon to get name + rank
    let detail = state.taxonomy.get_by_id(&id).await.unwrap_or(None);

    let (name, rank, kingdom) = match detail {
        Some(ref d) => (d.scientific_name.clone(), d.rank.clone(), d.kingdom.clone()),
        None => (id.clone(), "species".to_string(), None),
    };

    let options = TaxonOccurrenceOptions {
        limit: Some(limit),
        cursor: params.cursor,
        kingdom,
    };

    let rows = observing_db::feeds::get_occurrences_by_taxon(
        &state.pool,
        &name,
        &rank,
        &options,
        &state.hidden_dids,
    )
    .await?;

    let viewer = session_did(&cookies);
    let occurrences = enrichment::enrich_occurrences(
        &state.pool,
        &state.resolver,
        &state.taxonomy,
        &rows,
        viewer.as_deref(),
    )
    .await;

    let next_cursor = occurrences.last().map(|o| o.created_at.clone());

    Ok(Json(OccurrenceListResponse {
        occurrences,
        cursor: next_cursor,
    }))
}
